_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,220 @@
++  ArtifactID  uint      `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
++  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Total       uint      `gorm:"not null" json:"total"`
++  Failed      uint      `gorm:"not null" json:"failed"`
++  Passed      uint      `gorm:"not null" json:"passed"`
++  Skipped     uint      `gorm:"not null" json:"skipped"`
++}
+
++type TestResult struct {
++  ArtifactID  uint              `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact         `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Name        string	        `gorm:"not null" json:"name"`
++  Message     string            `gorm:"nullable" json:"message"`
++  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
++  Status      TestResultStatus  `gorm:"not null" enums:"Failed,Passed,Skipped"`
++}
+```
+```go
+// artifact.go
+// new
+type File struct {
+  name string
+  fileName string
+  data []bytes
+}
+// new, /build/{buildid}/test-results
+func (m ArtifactModule) putTestResultsHandler(c *gin.Context) {
+  // parseMultipartFormData to get the files
+  // for each file, parse and store in db
+}
+// new, /build/{buildid}/artifact/{artifactid}/test-results
+func (m ArtifactModule) getBuildArtifactTestResultsHandler(c *gin.Context) {
+  // fetch test results for the specified test
+}
+// new
+func parseMultipartFormData(c *gin.Context) []*File {
+  // ...
+}
+// new
+func parseTRX(file *File) []TestResult, TestResultSummary {
+  // ...
+}
+```
+
+The new endpoint `GET /build/{buildid}/test-results-summary` fetches
+all test result summaries for a build and constructs a json response
+that contains them, and has a summary of them, like:
+
+<details><summary>GET /build/42/test-results-summary</summary>
+
+```json
+{
+  "failed": 4,
+  "passed": 5,
+  "skipped": 0,
+  "artifacts": [
+    {
+      "artifactId": 123,
+      "filename": "MyResults1.trx",
+      "failed": 1,
+      "passed": 3,
+      "skipped": 0
+    },
+    {
+      "artifactId": 124,
+      "filename": "MyResults2.trx",
+      "failed": 3,
+      "passed": 2,
+      "skipped": 0
+    }
+  ]
+}
+```
+</details>
+
+### wharf-web
+
+wharf-web changes to use the new endpoints
+- `GET /build/{buildid}/test-results-summary`\
+  To get summary of all test result summaries for build.
+  
+- `GET /build/{buildid}/artifact/{artifactid}/test-results`\
+  To get test result details for build.
+  
+- `POST /build/{buildid}/test-results`\
+  To upload test result files for build.
```

Comment:
Maybe for unity with the naming difference, we should name this endpoint as `test-result-details` as well?

Reply:
Not entirely sure about this, since both details and summary get parsed (or generated, if the format uploaded doesn't include a summary like the current format does). Perhaps `POST /build/{buildid}/test-result-data`, as that is what's actually getting sent to the API.

Reply:
Yea, miss by me. In my head I pictured that the data you're uploading and downloading looks the same. But TRX isn't even close to our JSON format. Mb!